### ISSUE TITLE
[1.4.4.1] Build fails after upgrade, super not transpiled

### ISSUE THREAD

**Issue by dj-foxxy** [meteor/meteor#8578]
Hi all,

The following code fails to build after upgrading to 1.4.4.1 from 1.4.2.7:

```javascript
// @flow
import { Cell } from './Cell.js';

export class BatchNumberCell extends Cell {
  getRowClass(): string {
    const data = this._getRawData();

    if (data !== undefined && data !== null) {
      return 'line-item-started';
    }

    return super.getRowClass();
  }
}
```

With error message

```
While building for web.browser:
   client/templates/spreadsheet/cells/BatchNumberCell.js:47:15: 'super' outside of function or class (47:15)
```

The compiled code is:

```javascript
var BatchNumberCell = function (_Cell$Cell) {                                                                          //
  (0, _inherits3.default)(BatchNumberCell, _Cell$Cell);                                                                //
                                                                                                                       //
  function BatchNumberCell() {                                                                                         //
    (0, _classCallCheck3.default)(this, BatchNumberCell);                                                              //
    return (0, _possibleConstructorReturn3.default)(this, _Cell$Cell.apply(this, arguments));                          //
  }                                                                                                                    //
                                                                                                                       //
  BatchNumberCell.prototype.getRowClass = function () {                                                                //
    function getRowClass() {                                                                                           //
      var ret = function () {                                                                                          //
        var data = this._getRawData();                                                                                 //
                                                                                                                       //
        if (data !== undefined && data !== null) {                                                                     //
          return 'line-item-started';                                                                                  //
        }                                                                                                              // 2
                                                                                                                       //
        return super.getRowClass();                                                                                    //
      }.call(this);                                                                                                    //
                                                                                                                       //
      _assert(ret, _tcomb2.default.String, "return value");                                                            //
                                                                                                                       //
      return ret;                                                                                                      //
    }                                                                                                                  //
                                                                                                                       //
    return getRowClass;                                                                                                //
  }();                                                                                                                 //
                                                                                                                       //
  return BatchNumberCell;                                                                                              //
}(_Cell.Cell);                                                                                                         //
```

Looks like for some reason the `super` isn't getting transpiled.

If the flow return type is removed the code compiles correctly:

```javascript                                                                                                          //
var BatchNumberCell = function (_Cell$Cell) {                                                                          //
  (0, _inherits3.default)(BatchNumberCell, _Cell$Cell);                                                                //
                                                                                                                       //
  function BatchNumberCell() {                                                                                         //
    (0, _classCallCheck3.default)(this, BatchNumberCell);                                                              //
    return (0, _possibleConstructorReturn3.default)(this, _Cell$Cell.apply(this, arguments));                          //
  }                                                                                                                    //
                                                                                                                       //
  BatchNumberCell.prototype.getRowClass = function () {                                                                //
    function getRowClass() {                                                                                           //
      var data = this._getRawData();                                                                                   //
                                                                                                                       //
      if (data !== undefined && data !== null) {                                                                       //
        return 'line-item-started';                                                                                    //
      }                                                                                                                // 4
                                                                                                                       //
      return _Cell$Cell.prototype.getRowClass.call(this);                                                              // 6
    }                                                                                                                  //
                                                                                                                       //
    return getRowClass;                                                                                                //
  }();                                                                                                                 //
                                                                                                                       //
  return BatchNumberCell;                                                                                              //
}(_Cell.Cell);                                                                                                         //
```

**Comment by dj-foxxy** [comment]
Looks related to a transform in our `.babelrc`, namely `babel-tcomb`. So, might not be considered a bug with this project, Might be to do with babel upgrade. Feel free to close if you feel this is unrelated to Meteor. Still, thought it might be worth noting. Thanks.

**Comment by abernix** [comment]
Reproduction repo? :)

**Comment by dj-foxxy** [comment]
I knew you'd say that :) https://github.com/foxdog-studios/meteor-issue-8578

**Comment by benjamn** [comment]
@dj-foxxy As you may have gathered from https://github.com/meteor/meteor/issues/8422, `babel-plugin-tcomb` is a somewhat fragile beast.

In order to fix that issue, we made [this change](https://github.com/meteor/babel/commit/c30fe307fc1ce184f3aa8aa7712c87688bceb090) to force custom `.babelrc` plugins to run to completion before any other plugins or presets, which gives `babel-plugin-tcomb` a chance to do everything it needs before `babel-plugin-transform-es2015-modules-reify` starts visiting the AST.

Given that chance, however, it appears that `babel-plugin-tcomb` generates invalid code. For example, this class declaration:
```js
class DerivedClass extends BaseClass {
  foo(): number {
    return super.foo();
  }
}
```
becomes
```js
import _t from "tcomb";

class DerivedClass extends BaseClass {
  foo(): number {
    const ret = function () {
      return super.foo();
    }.call(this);

    _assert(ret, _t.Number, "return value");

    return ret;
  }
}
```
… if you run `babel-plugin-tcomb` and nothing else. This is invalid because the `super.foo` expression has been moved into a function expression, and is no longer in a class method.

Another way of interpreting this failure is that `babel-plugin-tcomb` is really hoping you're going to run `babel-plugin-transform-es2015-classes` in the same pass, to clean up the mess left behind. So, if you run
```sh
meteor npm i --save-dev babel-plugin-transform-es2015-classes
```
and change your `.babelrc` to
```js
{
  "plugins": [
    "tcomb",
    "transform-es2015-classes"
  ]
}
```
then you'll get
```js
var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

import _t from "tcomb";

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

let DerivedClass = function (_BaseClass) {
  _inherits(DerivedClass, _BaseClass);

  function DerivedClass() {
    _classCallCheck(this, DerivedClass);

    return _possibleConstructorReturn(this, (DerivedClass.__proto__ || Object.getPrototypeOf(DerivedClass)).apply(this, arguments));
  }

  _createClass(DerivedClass, [{
    key: "foo",
    value: function foo(): number {
      const ret = function () {
        return _get(DerivedClass.prototype.__proto__ || Object.getPrototypeOf(DerivedClass.prototype), "foo", this).call(this);
      }.call(this);

      _assert(ret, _t.Number, "return value");

      return ret;
    }
  }]);

  return DerivedClass;
}(BaseClass);
```
This is better because it's now valid code, but all those helpers are wasteful, so you'll probably also want to add `transform-runtime` to your `.babelrc`. Then you'll get:
```js
import _Object$getPrototypeOf from "babel-runtime/core-js/object/get-prototype-of";
import _possibleConstructorReturn from "babel-runtime/helpers/possibleConstructorReturn";
import _get from "babel-runtime/helpers/get";
import _inherits from "babel-runtime/helpers/inherits";
import _classCallCheck from "babel-runtime/helpers/classCallCheck";
import _createClass from "babel-runtime/helpers/createClass";
import _t from "tcomb";

let DerivedClass = function (_BaseClass) {
  _inherits(DerivedClass, _BaseClass);

  function DerivedClass() {
    _classCallCheck(this, DerivedClass);

    return _possibleConstructorReturn(this, (DerivedClass.__proto__ || _Object$getPrototypeOf(DerivedClass)).apply(this, arguments));
  }

  _createClass(DerivedClass, [{
    key: "foo",
    value: function foo(): number {
      const ret = function () {
        return _get(DerivedClass.prototype.__proto__ || _Object$getPrototypeOf(DerivedClass.prototype), "foo", this).call(this);
      }.call(this);

      _assert(ret, _t.Number, "return value");

      return ret;
    }
  }]);

  return DerivedClass;
}(BaseClass);
```
This should be enough to ensure `babel-preset-meteor` receives a well-formed syntax tree after your custom `.babelrc` plugins are finished running.

I realize that's a lot of information, but I hope it helps!

**Comment by dj-foxxy** [comment]
That's great thanks. I'll give it a go now. Also, thanks for taking the time to look at this issue. I know it's a corner case and the whole `babel-plugin-tcomb` thing is "carefully balanced". I always worry that my Github issues come across as "Fix this now!". But, especially in this case, it's more just seeing if something can be done easily.

*Backstory* (can be safely ignored)

Out of uni (5 years ago) my friend and I started a software shop. Our first gig was porting a manufacturing company's dBase 4 system to something more modern. We'd never done web dev before. We saw Meteor and realised it already took care of a lot of stuff we knew nothing about (I think we put it into production as soon as Meteor came out of beta). This project is still going on. So, not only is much of the code terrible but it's also like a Meteor museum: NPM wrapper packages, Blaze, React, PostgreSQL and MongoDB, NPM packages, CoffeeScript, ES5, ES2015+, .babelrc, Meteor packages to control load order, half the code in '/imports' half outside. Redux & Tracker, nested '`lib` directories to control load order, half Promise half Fibres, half modules half scripts.

We've got a lot better at web development and programming in general. We looked at the code base and realised what a mess we've made. This is were `tcomb` comes in. We're slowly creating a "green zone" of ES2015+ modules under `/imports` with strict type annotations. Then `babel-plugin-tcomb` acts as the border patrol. It has been a key tool when working with this codebase. It's allows us to slowly refactor and add new features to our old code, which often return a _surprising_ collections of types and structures.

Thanks.
